Step3 - 로또(자동)

livedetuc opened the pull request:
step2에 꼼꼼하게 리뷰 해주셔서 감사했습니다!! :) 👍  (최고)

로또 구현했는데, class를 너무 세세히 나누지는 않았나 모르겠네요.
불금에 리뷰 요청 드리는 저를 용서하세요

감사합니다~!

Commit:
Step2 - 리뷰 피드백 반영

- 매직넘버 사용 지양
- text, str 등의 의미가 모호한 파라미터명 변경
- 변하지 않는 field 는 final 처리
- util 성 class에는 private 생성자 추가
- @NullAndEmptySource 사용하여 test case 통합
- 중복 validate 호출 부분 삭제

Commit:
doc (README.md) - 3단계 로또(자동) 기능 목록 정의

Commit:
test, feat (LottoNumber) - 로또 숫자 domain, test 구현

Commit:
test, feat (LottoTicket) - 로또 티켓 domain, test 구현

Commit:
test, feat (LottoTickets) - 로또 티켓 목록 domain, test 구현

Commit:
test, feat (LottoMatchResult) - 로또 숫자 매치 결과 domain, test 구현

- lottoTicket: 번호 입력받는(당첨번호) 생성자와 테스트 추가
- lottoTickets: 티켓 목록 입력 받는 생성자 추가

Commit:
test, feat (LottoTicketsPrice) - 로또 구입 금액 domain, test 구현

Commit:
test, feat (LottoWinningPrice) - 로또 당첨금 domain, test 구현

Commit:
test, feat (LottoWinningProfit) - 로또 수익률 domain, test 구현

- view: InputView, ResultView 구현
- ManualLottoApplication 실행 코드 구현

nokchax commented on src/main/java/step3/ManualLottoApplication.java:
```diff
@@ -0,0 +1,40 @@
+    LottoTicket winningTicket = new LottoTicket(InputView.getWinningNumbersInput());
+
+    // 당첨 결과 출력
+    LottoMatchResult lottoMatchResult = new LottoMatchResult(lottoTickets,
+        winningTicket); // 숫자 매칭 결과
```
실수로 개행하신걸까요?
```suggestion
    LottoMatchResult lottoMatchResult = new LottoMatchResult(lottoTickets, winningTicket); // 숫자 매칭 결과
```

livedetuc replied:
java 구글 코드스타일 xml을 넣고 reformat했더니 자동 개행 되었습니다~!

nokchax commented on src/main/java/step3/domain/LottoNumber.java:
```diff
@@ -0,0 +1,25 @@
+
+  public LottoNumber(int number) {
+    this.number = number;
+    validate();
```
사소한 건데, 유효성 검은은 값 할당 전에 해야하지 않을까요?

nokchax commented on src/main/java/step3/domain/LottoTicket.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  public List<Integer> getNumbersAsInteger() {
+    return this.numbers.stream().map(LottoNumber::getNumber).collect(Collectors.toList());
```
```suggestion
    return this.numbers.stream()
                      .map(LottoNumber::getNumber)
                      .collect(Collectors.toList());
```
가독성을 위해 개행을 적절히 섞어보는건 어떨까요?

nokchax commented on src/main/java/step3/domain/LottoTicket.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  private List<Integer> getTotalNumbers() {
+    final List<Integer> totalNumbers = new ArrayList<>();
+    for (int i = LottoNumber.MIN_NUMBER; i <= LottoNumber.MAX_NUMBER; i++) {
+      totalNumbers.add(i);
+    }
+    return totalNumbers;
```
전반적으로 for문과 스트림을 혼용하시는데, 참고해보세요
```suggestion
    return IntStream.rangeClosed(LottoNumber.MIN_NUMBER, LottoNumber.MAX_NUMBER)
            .boxed()
            .collect(Collectors.toList());
```

nokchax commented on src/main/java/step3/domain/LottoTicket.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  private List<LottoNumber> generateRandomLottoNumbers() {
+    final List<Integer> totalNumbers = getTotalNumbers();
+    Collections.shuffle(totalNumbers);
+
+    final List<Integer> lottoNumbers = totalNumbers.subList(0, NUMBERS_COUNT);
+    Collections.sort(lottoNumbers);
+
+    return lottoNumbers.stream()
+        .map(LottoNumber::new)
+        .collect(Collectors.toList());
```
로또는 1~45의 정해진 범위 내에서만 생성되는데요, 
이 LottoNumber 객체를 미리 생성해서 캐시하고
로또 티켓을 생성할때 가져다가 사용하는건 어떨까요?

nokchax commented on src/main/java/step3/domain/LottoTicket.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  private void validate() {
+    if (this.numbers == null || this.numbers.size() != NUMBERS_COUNT
+        || this.numbers.stream().distinct().count() != NUMBERS_COUNT) {
```
`this.numbers.size() != NUMBERS_COUNT` 는 아래의 로직으로 같이 걸러지지 않을까요? 
`this.numbers.stream().distinct().count() != NUMBERS_COUNT`

livedetuc replied:
'1, 1, 2, 3, 4, 5, 6' 같은 numbers가 들어왔을 케이스를 생각해 numbers.size를 따로 체크했습니다

nokchax replied:
네 '1,2,3,4,5,6' 과 같은 케이스도 distinct().count()로 걸러지니 중복일거 같아서요

livedetuc replied:
'1, 1, 2, 3, 4, 5, 6'  는 disctinct().count()는 6이고, .count() 는 7 인데,
잘못된 인풋이라고 생각해서 .count()를 통해 거르고자 했습니다~! :)

nokchax commented on src/main/java/step3/domain/LottoTicket.java:
```diff
@@ -0,0 +1,65 @@
+    }
+  }
+
+  @Override
+  public String toString() {
+    return this.numbers.stream().map(LottoNumber::getNumber).collect(Collectors.toList())
+        .toString();
+  }
```
전반적으로 도메인 객체에서 toString을 많이 사용하셨는데요
도메인 객체내에서 view와 관련된 로직이 있으면 MVC구조에 맞다고 볼 수 있을까요?
DTO를 활용해 보면 어떨까요

livedetuc replied:
넵 view로 옮겨보겠습니다~

nokchax commented on src/main/java/step3/domain/LottoTickets.java:
```diff
@@ -0,0 +1,31 @@
+    this.lottoTickets = lottoTicketList;
+  }
+
+  public List<LottoTicket> getLottoTickets() {
+    return lottoTickets;
+  }
```
전반적으로 getter 사용도 많이 하셨는데, 클래스를 나누는 것 만으로는 객체지향 프로그래밍이라고 할 수 없습니다.
클래스를 나누고 클래스간에 메세지를 전달해서 새로운 객체를 만들어내거나 상태를 변경하는 방식을 적용해보세요 :)

e.g.
```java
LottoMatchResult lottoMatchResult = new LottoMatchResult(lottoTickets, winningTicket); // 숫자 매칭 결과
```
대신에
```java
LottoMatchResult lottoMatchResult = lottoTickets.matchWinningNumbers(winningTicket); // 숫자 매칭 결과
```

nokchax commented on src/main/java/step3/domain/LottoWinningPrice.java:
```diff
@@ -0,0 +1,50 @@
+
+  // key: matchCount, value: winningPrice
+  static final public Map<Integer, Integer> MATCH_COUNT_WINNING_PRICE_MAP = new HashMap<Integer, Integer>() {{
+    put(3, 5_000);
+    put(4, 50_000);
+    put(5, 1_500_000);
+    put(6, 2_000_000_000);
```
이 매치 카운트와 상금은 enum으로 추출할 수 있지 않을까요?

nokchax commented on src/main/java/step3/domain/LottoTicketsPrice.java:
```diff
@@ -0,0 +1,24 @@
+public class LottoTicketsPrice {
+
+  private final int UNIT_TICKET_PRICE = 1000;
+  private int purchasePrice;
```
price 도 클래스로 추출할 수 있지 않을까요?

nokchax replied:
LottoTicketsPrice 와 LottoWinningPrice 내의 price를 Money 클래스로 추출하면
각종 연산에 대한 로직을 모으고 중복을 제거할 수 있을거 같아요

nokchax reviewed (changes requested):
안녕하세요 한결님
세번째도 과제 진행하느라 고생하셨습니다 👍 
코멘트 남겨봤는데요, 확인 후에 재요청 부탁드립니다 🙇‍♂️

Commit:
fix - 리뷰어님 피드백 반영

- 객체 출력 시 domain.toString() 대신에 view 에서 로직을 구현하도록 이동
- 1~45 사이 숫자 목록을 미리 생성하여 로또 티켓 생성시 가져다 쓰도록 변경 (LottoTicket.totalLottoNumbers)
- for문과 스트림을 혼용하지 않도록 스트림으로 변경
- 생성자에서 값 할당 전에 유효값 검증(validate) 하도록 순서 변경
- 매치카운트와 상금을 enum으로 추출

livedetuc commented:
리뷰 주신 내용 추가로 작업해보았습니다~!
확인 부탁드립니다! 감사합니다 :)

nokchax commented on src/main/java/step3/domain/LottoMatchCaseEnum.java:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  public static LottoMatchCaseEnum value(int matchCount) {
+    switch (matchCount) {
+      case 0:
+        return ZERO_NUMBER_MATCH;
+      case 1:
+        return ONE_NUMBER_MATCH;
+      case 2:
+        return TWO_NUMBERS_MATCH;
+      case 3:
+        return THREE_NUMBERS_MATCH;
+      case 4:
+        return FOUR_NUMBERS_MATCH;
+      case 5:
+        return FIVE_NUMBERS_MATCH;
+      case 6:
+        return SIX_NUMBERS_MATCH;
+      default:
+        return null;
+    }
```
switch 문도 if else 문이라고 봐야해요
`규칙 2: else 예약어를 쓰지 않는다.` 를 지켜주세요!
for문이나 스트림을 사용해서도 카운트에 따라 매칭되는 enum을 찾는 방법이 있답니다.

nokchax commented on src/main/java/step3/domain/LottoMatchResult.java:
```diff
@@ -1,34 +1,31 @@
-  static final public int MATCH_COUNT_MAX = 6;
-
-  final private Map<Integer, Integer> matchCountMap = new HashMap<>();
+  final private Map<LottoMatchCaseEnum, Integer> matchCountMap = new HashMap<>();
```
👍

nokchax commented on src/main/java/step3/domain/LottoNumber.java:
```diff
@@ -1,19 +1,26 @@
-
   private final int number;
 
+  @Override
+  public int compareTo(LottoNumber o) {
+    if (this.number > o.getNumber()) {
+      return 1;
+    }
+    return -1;
+  }
+
```
일반적인 메서드는 생성자 아래에 추가해 주세요!
```
class A {
    상수(static final) 또는 클래스 변수
    인스턴스 변수
    생성자
    팩토리 메소드
    메소드
    기본 메소드 (equals, hashCode, toString)
}
```

nokchax commented on src/main/java/step3/domain/LottoTicket.java:
```diff
@@ -18,48 +24,40 @@ public LottoTicket() {
   @Override
   public String toString() {
-    return this.numbers.stream().map(LottoNumber::getNumber).collect(Collectors.toList())
+    return this.numbers.stream()
+        .map(LottoNumber::getNumber)
+        .collect(Collectors.toList())
```
toString 이 아직 남아있습니다~! 도메인에서 뷰와 관련된 로직을 분리해 주세요 :)

nokchax reviewed (changes requested):
피드백 반영 잘 해주셨습니다 👍 
아직 반영이 덜 된 부분이 보여서 코멘트 남겼는데 확인 후 재요청 부탁드립니다 🙇‍♂️ 💪

Commit:
fix - 리뷰어님 피드백 반영 2

- 객체 출력 시 domain.toString() 대신에 view 에서 로직을 구현하도록 이동 (LottoTicket)
- swich-case 문 대신 stream을 이용하도록 변경 (LottoMatchCaseEnum)
- 일반 메소드가 생성자 다음에 오도록 위치 변경 (LottoNumber)
- LottoTicketsPrice, LottoWinningPrice 두 클래스를 Money 클래스로 추출

livedetuc commented:
두번째 피드백도 반영해보았습니다 ~!

LottoTicketsPrice 와 LottoWinningPrice 를 money로 통합하였는데,
다시 읽어보니 class내의 price 필드만 추출&통합을 의미하셨던 걸까요?

확인해주시고 리뷰 다시 부탁드립니다 :)
감사합니다!!

nokchax commented on src/main/java/step3/domain/LottoMatchCaseEnum.java:
```diff
@@ -18,24 +20,10 @@
-      default:
-        return null;
-    }
+    return Arrays.stream(LottoMatchCaseEnum.values())
+        .filter(matchCaseEnum -> matchCaseEnum.getMatchCount() == matchCount)
+        .findAny()
+        .orElse(null);
```
👍👍
다만 null를 리턴하는 거는 삼가해주세요, 예외적인 상황이라 봐야할것 같으니 Exception을 던지는게 낫겠네요 :)
클린코드 7장을 보시면 null에 관한 이야기들이 나오니 참고삼아 읽어보세요~!

nokchax commented on src/main/java/step3/domain/LottoWinningProfit.java:
```diff
@@ -15,7 +15,8 @@ public float getProfit() {
   private void validate(float profit) {
     if (profit < 0) {
-      throw new RuntimeException("[ERROR] winning price profit cannot be negative. profit =" + profit);
+      throw new RuntimeException(
```
RuntimeException 은 너무 상위의 예외입니다. 
지나치게 포괄적이기 때문에 적당한 예외를 사용하는 걸 추천드릴게요, 
이와같은 경우는 예외적인 인풋값이니까 `IlleagalArgumentException`이 어울리겠네요

nokchax commented on src/main/java/step3/domain/Money.java:
```diff
@@ -0,0 +1,35 @@
+
+public class Money {
+
+  private final int UNIT_TICKET_PRICE = 1000;
```
상수는 `final` 키워드도 잊지 마세요~!

livedetuc replied:
final 키워드 들어가 있습니다~

nokchax commented on src/main/java/step3/domain/Money.java:
```diff
@@ -0,0 +1,35 @@
+    }
+  }
+
+  public int getTicketCount() {
+    return (int) (this.price / UNIT_TICKET_PRICE);
+  }
```
메소드 추출 👍  메소드명은 Money라는 객체와 조금 안어울리긴 하네요 다음단계에서 한번 고민해주세요!

nokchax replied:
질문 주신걸 리뷰 끝나고 나서 봤네요
전에 남긴 코멘트는 Money에 대한 추출을 의미한거였습니다.

nokchax commented on src/main/java/step3/view/ResultView.java:
```diff
@@ -4,15 +4,20 @@
+  }
+
+  private static void printTicket(LottoTicket lottoTicket) {
+    System.out.println(lottoTicket.getNumbers().stream()
+        .map(LottoNumber::getNumber)
+        .collect(Collectors.toList()));
```
도메인에서 view와 관련된 로직을 잘 분리해 내셨습니다 💯 
다음에는 dto 사용도 도전해 보세요!

nokchax reviewed (approved):
안녕하세요, 한결님. 피드백 잘 반영해주셨습니다 👍
이번 단계는 이만 머지하겠습니다, 몇몇 코멘트는 다음 단계에서 고민해주세요 :)
